Submitting a modular and automatic Config API

Poslovitch opened the pull request:
a lot of unused/useless Settings has been removed
Settings class has been moved to the main package, because the config package would then only contain it

I need feedback about it

Poslovitch committed:
ConfigAPI - Implementing the basic functionnalities

a lot of unused/useless Settings has been removed
Settings class has been moved to the main package, because the config package would then only contain it

I need feedback about it

tastybento committed:
Extended the config api annotations.

Added the concept of an adapter class that would convert the YAML input
to the setting type and value.
I included an example adapter class for Enum.

I fixed a number of the settings to match the config/yml or the other
way around.

After compiling and running, do /bsb reload to see the result.

This is a WIP and obviously not finished.

Poslovitch commented on src/main/java/us/tastybento/bskyblock/config/ConfigLoader.java:
```diff
@@ -0,0 +1,69 @@
+                            }
+                        } else if (configValue != null){
+                            // No adapter - I hope this works!
+                            field.set(null, configValue);
```
I don't understand the `field.set(null, configValue);`

Why do you apply the configValue to a null object?

tastybento replied:
This has to be done if the field is a static - i.e., it has no instance behind it. In Settings, all the fields are statics and the Settings class is never instanced by BSkyBlock. There are some programmers who don't like this, and we could change it. See my next comment.

tastybento committed:
Added annotation handling to the flatfile database handler.

Currently it handles the @ConfigEntry path and specificTo fields.

Experimental: There is a class called Setting2.java. It has an
annotation that defines its filename as config.yml. Currently, it is
using the database folder as its location, but it could be the plugin's
datafolder in the future. By placing the config.yml file in the database
folder, it will be read. See the code in BSkyBlock onEnable() for how
that is done.

The main differences between Settings2.java and Settings.java are that
all fields are NOT static and therefore it is an object and uses getters
and setters. This is because it is a JavaBean. In other code, settings
should be queried using this config object.

tastybento committed:
Converted to using the Settings loaded by the database handler.

Removed a lot of static references.

Constants have their own class.

tastybento committed:
Fixed code according to comments on last commit.

Reworked locales so they use a nice and pretty folder structure.

Added the ability to define adapters to serialize certain data
structures. Added the PotionType list adapter.

Saving and loading of configs basically works. Known issues:

1. The config.yml is saved and loaded to/from the database folder
2. More error handling is required for the config loading. e.g., a list
value with only one value in it is not read as a list. This could get
tricky.
3. Comments are not saved (yet)

tastybento committed:
Added saveSettings and loadSettings methods.

These methods are used specifically for loading and saving settings
classes (those use the ISettings interface). The saving saves a copy of
any settings class in the database for future reference. The loading
loads a copy from the database if it exists and checks if any fields are
different from the config file. If they are different and some action
needs to be taken, the action is taken. This is still be to be coded.

In other news, the saving of arrays is unsupported and currently should
be avoided. Use Lists or Sets instead.

tastybento committed:
Now uses the official config.yml in the BSkyBlock data folder.

Saves a backup of the config after validation in the database so it can
be checked.

To Be Done: Validation of the config when it is loaded against the
database version.

tastybento commented on src/main/java/us/tastybento/bskyblock/database/flatfile/FlatFileDatabaseHandler.java:
```diff
@@ -379,18 +484,30 @@ private Object deserialize(Object value, Class<? extends Object> clazz) {
+    public T loadSettings(String uniqueId, T dbConfig) throws InstantiationException, IllegalAccessException, IllegalArgumentException, InvocationTargetException, ClassNotFoundException, IntrospectionException {
+        if (dbConfig == null) return loadObject(uniqueId);
+        
+        // TODO: compare the loaded with the database copy
```
This is where the annotations needsReset = true and overrideOnChange = true need to be processed.
